File: android/data/repositories/src/main/kotlin/com/gemwallet/android/data/repositories/gemstone/BalanceStore.kt
package com.gemwallet.android.data.repositories.gemstone

import com.gemwallet.android.data.service.store.database.BalancesDao
import kotlinx.coroutines.flow.first
import com.gemwallet.android.data.service.store.database.AssetsDao
import com.gemwallet.android.data.service.store.database.entities.DbBalance
import com.gemwallet.android.serializer.decodeJson
import com.wallet.core.primitives.BalanceMetadata
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import uniffi.gemstone.GemBalanceStore
import uniffi.gemstone.GemBalanceUpdate
import uniffi.gemstone.GemBalanceUpdateType

class GemstoneBalanceStore(
    private val balancesDao: BalancesDao,
    private val assetsDao: AssetsDao,
) : GemBalanceStore {

    override suspend fun getEnabledAssetIds(walletId: String, assetIds: List<String>): List<String> =
        assetsDao.getAssetsInfo(walletId, assetIds).first().filter { it.visible == true }.map { it.id }

    override suspend fun setEnabled(walletId: String, assetIds: List<String>, enabled: Boolean) {
        assetIds.forEach { assetsDao.setWalletAssetVisibility(walletId, it, enabled) }
    }

    override suspend fun setPinned(walletId: String, assetId: String, pinned: Boolean) {
        val balance = assetsDao.getBalance(walletId, assetId) ?: return
        assetsDao.setBalanceConfig(walletId, assetId, isPinned = pinned, isVisible = balance.isVisible, listPosition = balance.listPosition)
    }

    override suspend fun updateBalances(walletId: String, updates: List<GemBalanceUpdate>) = withContext(Dispatchers.IO) {
        val updatedAt = System.currentTimeMillis()
        for (update in updates) {
            balancesDao.insertIgnore(DbBalance(assetId = update.assetId, walletId = walletId, isVisible = false, updatedAt = updatedAt))
            when (val type = update.updateType) {
                is GemBalanceUpdateType.Coin -> balancesDao.updateCoinBalance(
                    walletId = walletId,
                    assetId = update.assetId,
                    available = type.available.value,
                    availableAmount = type.available.amount,
                    reserved = type.reserved.value,
                    reservedAmount = type.reserved.amount,
                    pendingUnconfirmed = type.pendingUnconfirmed.value,
                    pendingUnconfirmedAmount = type.pendingUnconfirmed.amount,
                    isActive = update.isActive,
                    updatedAt = updatedAt,
                )
                is GemBalanceUpdateType.Token -> balancesDao.updateTokenBalance(
                    walletId = walletId,
                    assetId = update.assetId,
                    available = type.available.value,
                    availableAmount = type.available.amount,
                    isActive = update.isActive,
                    updatedAt = updatedAt,
                )
                is GemBalanceUpdateType.Stake -> {
                    val metadata = type.metadata?.decodeJson<BalanceMetadata>()
                    balancesDao.updateStakeBalance(
                        walletId = walletId,
                        assetId = update.assetId,
                        staked = type.staked.value,
                        stakedAmount = type.staked.amount,
                        frozen = type.frozen.value,
                        frozenAmount = type.frozen.amount,
                        locked = type.locked.value,
                        lockedAmount = type.locked.amount,
                        pending = type.pending.value,
                        pendingAmount = type.pending.amount,
                        rewards = type.rewards.value,
                        rewardsAmount = type.rewards.amount,
                        votes = metadata?.votes?.toLong() ?: 0L,
                        energyAvailable = metadata?.energyAvailable?.toLong() ?: 0L,
                        energyTotal = metadata?.energyTotal?.toLong() ?: 0L,
                        bandwidthAvailable = metadata?.bandwidthAvailable?.toLong() ?: 0L,
                        bandwidthTotal = metadata?.bandwidthTotal?.toLong() ?: 0L,
                        updatedAt = updatedAt,
                    )
                }
                is GemBalanceUpdateType.Perpetual -> balancesDao.updatePerpetualBalance(
                    walletId = walletId,
                    assetId = update.assetId,
                    available = type.available.value,
                    availableAmount = type.available.amount,
                    reserved = type.reserved.value,
                    reservedAmount = type.reserved.amount,
                    withdrawable = type.withdrawable.value,
                    withdrawableAmount = type.withdrawable.amount,
                    isActive = update.isActive,
                    updatedAt = updatedAt,
                )
                is GemBalanceUpdateType.Earn -> balancesDao.updateEarnBalance(
                    walletId = walletId,
                    assetId = update.assetId,
                    earn = type.balance.value,
                    earnAmount = type.balance.amount,
                    isActive = update.isActive,
                    updatedAt = updatedAt,
                )
            }
        }
    }
}

File: ios/Packages/GemstoneServices/Sources/Stores/BalanceStore.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import protocol Gemstone.GemBalanceStore
import struct Gemstone.GemAssetBalance
import struct Gemstone.GemBalanceRecord
import struct Gemstone.GemBalanceValue
import typealias Gemstone.AssetId
import GemstonePrimitives
import Primitives
import Store

public final class GemstoneBalanceStore: GemBalanceStore, @unchecked Sendable {
    private let store: BalanceStore

    public init(store: BalanceStore) {
        self.store = store
    }

    public func getAvailableBalances(walletId: String, assetIds: [Gemstone.AssetId]) throws -> [GemAssetBalance] {
        let walletId = try WalletId.from(id: walletId)
        return try assetIds.compactMap { assetId in
            let assetId = try Primitives.AssetId(id: assetId)
            return try store.getBalance(walletId: walletId, assetId: assetId).map { GemAssetBalance($0.balance, assetId: assetId, isActive: $0.isActive) }
        }
    }

    public func updateBalances(walletId: String, balances: [GemBalanceRecord]) async throws {
        let walletId = try WalletId.from(id: walletId)
        let updates = try balances.map { balance in
            try UpdateBalance(
                assetId: Primitives.AssetId(id: balance.assetId),
                available: value(balance.available),
                frozen: value(balance.frozen),
                locked: value(balance.locked),
                staked: value(balance.staked),
                pending: value(balance.pending),
                pendingUnconfirmed: value(balance.pendingUnconfirmed),
                rewards: value(balance.rewards),
                reserved: value(balance.reserved),
                withdrawable: value(balance.withdrawable),
                earn: value(balance.earn),
                metadata: balance.metadata.map { $0.map() },
                updatedAt: .now,
                isActive: balance.isActive,
            )
        }
        try store.updateBalances(updates, for: walletId)
    }

    public func getEnabledAssetIds(walletId: String) async throws -> [Gemstone.AssetId] {
        try store.getEnabledAssetIds(walletId: WalletId.from(id: walletId)).map(\Primitives.AssetId.identifier)
    }

    public func setAssetsEnabled(walletId: String, assetIds: [Gemstone.AssetId], enabled: Bool) async throws {
        try store.setIsEnabled(walletId: WalletId.from(id: walletId), assetIds: assetIds.map { try Primitives.AssetId(id: $0) }, value: enabled)
    }

    public func setAssetPinned(walletId: String, assetId: Gemstone.AssetId, pinned: Bool) async throws {
        try store.pinAsset(walletId: WalletId.from(id: walletId), assetId: Primitives.AssetId(id: assetId), value: pinned)
    }

    private func value(_ value: GemBalanceValue) -> UpdateBalanceValue {
        UpdateBalanceValue(value: value.value.description, amount: value.amount)
    }
}
